### PR TITLE
dnat to 169.254.169.252 for metadata server traffic

### PR DESCRIPTION
**Pull request by zshihang**
#### What type of PR is this?

/kind regression

#### What this PR does / why we need it:

REDIRECT doesn't work well with calico clusters.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -173,8 +173,9 @@ function config-ip-firewall {
   # node because we don't expect the daemonset to run on this node.
   if [[ "${ENABLE_METADATA_CONCEALMENT:-}" == "true" ]] && [[ ! "${METADATA_CONCEALMENT_NO_FIREWALL:-}" == "true" ]]; then
     echo "Add rule for metadata concealment"
-    iptables -w -t nat -I PREROUTING -p tcp ! -i eth0 -d "${METADATA_SERVER_IP}" --dport 80 -m comment --comment "metadata-concealment: bridge traffic to metadata server goes to metadata proxy" -j REDIRECT --to-ports 988
-    iptables -w -t nat -I PREROUTING -p tcp ! -i eth0 -d "${METADATA_SERVER_IP}" --dport 8080 -m comment --comment "metadata-concealment: bridge traffic to metadata server goes to metadata proxy" -j REDIRECT --to-ports 987
+    ip route add table local local 169.254.169.252 dev lo proto kernel scope host src 169.254.169.252
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh:176]: Can we just add this address to eth0?
```
ip addr add dev eth0 169.254.169.252/32
```
It's a lot more obvious then hiding something in the local route table.

Comment (zshihang) [reply cluster/gce/gci/configure-helper.sh:176]: done

Comment (aojea) [reply cluster/gce/gci/configure-helper.sh]: you can add it to the loopback interface, any special reason to use eth0? 
```
$ sudo ip addr add 169.254.169.252 dev lo
$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet 169.254.169.252/32 scope global lo
```
is not a good idea to have this link local addresses in public interfaces ... but is not wrong though

Comment (zshihang) [reply cluster/gce/gci/configure-helper.sh]: thanks for pointing it out. i meant to add it to lo. done.

Comment (mikedanese) [reply cluster/gce/gci/configure-helper.sh]: Apparently route_localnet only cares about 127/8, not that the requests target loopback?

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv4/route.c?h=v5.11#n1710

If this works in testing, lo seems preferable to eth0.

Comment (aojea) [reply cluster/gce/gci/configure-helper.sh]: yeah, localnet is hack to solve the martian problem with localhost addresses 
```
route_localnet - BOOLEAN
    Do not consider loopback addresses as martian source or destination
    while routing. This enables the use of 127/8 for local routing purposes.
    default FALSE
```

this is just another ip ... in some other interface

**Comment by hermanbanken** [comment]
Is it possible that this was in GKE 1.21.5 and that it changes the source address for the health checks to be 169.x.x.x?

We're observing that behavior and had local subnet ips whitelisted for health checks, but not this one. So the pods became unhealthy/unlively.

**Comment by zshihang** [comment]
this change is in gke 1.21. see https://cloud.google.com/kubernetes-engine/docs/release-notes#October_28_2021.

**Comment by hermanbanken** [comment]
This only talks about the Metadata API but it was not clear to me that this involves the health checks too. Is that just assumed knowledge?

**Comment by zshihang** [comment]
are you talking about the healthz endpoint of gke metadata server? there is no change and you can still curl the endpoints using `curl localhost:989/healthz` on the node vm.

other health checks should not be affected by this change as well.

**Comment by hermanbanken** [comment]
Sorry for the confusion, I mean the readiness and liveness probes

**Comment by zshihang** [comment]
`169.254.169.252` was added in `global` scope. due to https://github.com/kubernetes/kubernetes/pull/95790, kubelet would select it as source ip in some cases instead of node ip. 

we are going to address it in https://github.com/kubernetes/kubernetes/pull/107502.

**Pull request merged**